src/storage/task.ts

```diff
@@ -0,0 +1,42 @@
+import { storage } from 'firebase/app';
+import { fromTask } from './observable/fromTask';
+import { Observable } from 'rxjs/Observable';
+import { map, filter, shareReplay } from 'rxjs/operators';
+
+export interface AngularFireUploadTask {
+  snapshotChanges(): Observable<storage.UploadTaskSnapshot | undefined>;
+  percentageChanges(): Observable<number | undefined>;
+  downloadURL(): Observable<string | null>;
+  pause(): boolean;
+  cancel(): boolean;
+  resume(): boolean;
+  then(): Promise<any>;
+  catch(onRejected: (a: Error) => any): Promise<any>;
+}
+
+export function createUploadTask(task: storage.UploadTask): AngularFireUploadTask {
+  const inner$ = fromTask(task).pipe(shareReplay());
```

Comment:
Given that `fromTask(task)` is going to return a multicast observable by default (since it doesn't create the task, rather, it wraps it), you probably don't need the `shareReplay` unless you're trying to get a caching behavior. What is the use case where you will need that caching behavior? To be specific, you're talking about a scenario where the fromTask observable completes, but you want to subscribe to it later and get the value(s) out of it.

Reply:
Since this is a file upload upload, I'd like to keep from uploading it again. If a user subscribes multiple times it be ideal to replay the state rather than upload the file again.

Reply:
You might want to let your users compose that in. Otherwise, if they want to remove caching, they might have a hard time. (A `take(1)` should do it) But you'll be taking away from them a primitive building block. 

A use case might help me understand though.